src/tabmat/constructor.py

```diff
@@ -133,53 +133,205 @@ def from_pandas(
+        Name of the category to which missing values will be converted if
+        ``cat_missing_method='convert'``.
+
+    Returns
+    -------
+    SplitMatrix
+    """
+    matrices: list[Union[DenseMatrix, SparseMatrix, CategoricalMatrix]] = []
+    indices: list[list[int]] = []
+    is_cat: list[bool] = []
+
+    dense_columns = []  # column index in original DataFrame
+    dense_indices = []  # index in the new SplitMatrix
+    sparse_columns = []  # sparse columns to join together
+    sparse_indices = []  # index in the new SplitMatrix
+    ignored_cols = []
+
+    mxcolidx = 0
+
+    for coldata in df.iter_columns():
+        if isinstance(coldata.dtype, (pl.Categorical, pl.Enum)):
+            cat = CategoricalMatrix(
+                coldata,
+                drop_first=drop_first,
+                dtype=dtype,
+                column_name=coldata.name,
+                term_name=coldata.name,
+                column_name_format=categorical_format,
+                cat_missing_method=cat_missing_method,
+                cat_missing_name=cat_missing_name,
+            )
+            if len(cat.categories) < cat_threshold:
+                (
+                    X_dense_F,
+                    X_sparse,
+                    dense_idx,
+                    sparse_idx,
+                ) = _split_sparse_and_dense_parts(
+                    sps.csc_matrix(cat.tocsr(), dtype=dtype),
+                    threshold=sparse_threshold,
+                    column_names=cat.get_names("column"),
+                    term_names=cat.get_names("term"),
+                )
+                matrices.append(X_dense_F)
+                is_cat.append(True)
+                matrices.append(X_sparse)
+                is_cat.append(True)
+                if cat_position == "expand":
+                    indices.append(mxcolidx + dense_idx)
+                    indices.append(mxcolidx + sparse_idx)
+                    mxcolidx += len(dense_idx) + len(sparse_idx)
+                elif cat_position == "end":
+                    indices.append(dense_idx)
+                    indices.append(sparse_idx)
+
+            else:
+                matrices.append(cat)
+                is_cat.append(True)
+                if cat_position == "expand":
+                    indices.append(mxcolidx + np.arange(cat.shape[1]))
+                    mxcolidx += cat.shape[1]
+                elif cat_position == "end":
+                    indices.append(np.arange(cat.shape[1]))
+        elif coldata.dtype.is_numeric():
+            if (coldata != 0).mean() <= sparse_threshold:
+                sparse_columns.append(coldata.name)
+                sparse_indices.append(mxcolidx)
+                mxcolidx += 1
+            else:
+                dense_columns.append(coldata.name)
+                dense_indices.append(mxcolidx)
+                mxcolidx += 1
+
+        else:
+            ignored_cols.append(coldata.name)
+
+    if len(ignored_cols) > 0:
```

Comment:
This later part looks very similar to the Pandas version. Might we be able to ducktype this with a shared helper function?

Reply:
Happy to add a helper function, but is it worth it? It'd require lots of inputs and would either modify them in place or return lots of outputs, both of which make the code less transparent.

Reply:
Would it be so many inputs and outputs? Can't the helper function just return the single matrix at the end? I'd say it is fine for a private helper function to modify the inputs if it significantly cuts down on the duplicated code.